pkg/sfu/receiver.go

```diff
@@ -360,52 +334,80 @@ func (w *WebRTCReceiver) writeRTP(layer int) {
 			return
 		}
 
-		w.locks[layer].Lock()
+		if w.isSimulcast {
+			pts := w.pendingTracks[layer].Load().([]*DownTrack)
+			if len(pts) > 0 {
+				if pkt.KeyFrame {
+					w.Lock()
+					for _, dt := range pts {
+						w.DeleteDownTrack(dt.CurrentSpatialLayer(), dt.peerID)
+						w.storeDownTrack(layer, dt)
+						dt.SwitchSpatialLayerDone(int32(layer))
 
-		if w.isSimulcast && len(w.pendingTracks[layer]) > 0 {
-			if pkt.KeyFrame {
-				for _, dt := range w.pendingTracks[layer] {
-					w.downTracks[layer] = append(w.downTracks[layer], dt)
-					w.DeleteDownTrack(dt.CurrentSpatialLayer(), dt.peerID)
-					dt.SwitchSpatialLayerDone()
+					}
+					w.pendingTracks[layer].Store(make([]*DownTrack, 0, 10))
+					w.Unlock()
+				} else {
+					w.SendRTCP(pli)
 				}
-				w.pendingTracks[layer] = w.pendingTracks[layer][:0]
-			} else {
-				w.SendRTCP(pli)
 			}
 		}
 
-		for _, dt := range w.downTracks[layer] {
-			if err = dt.WriteRTP(pkt); err != nil {
+		dts := w.downTracks[layer].Load().([]*DownTrack)
+		for _, dt := range dts {
+			if err = dt.WriteRTP(pkt, layer); err != nil {
+				if err == io.EOF && err == io.ErrClosedPipe {
+					w.Lock()
```

Comment:
I think we perform this in several places, Lock then DeleteDownTrack. should the lock/unlock be moved into DeleteDownTrack?

Reply:
Originally was like that, but on layer change we want to perform many operations, that means that we would need to get a lock and free it 3 times.